ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldStatePeerTrieNodeFinder.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright contributors to Hyperledger Besu
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.sync.worldstate;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.snap.RetryingGetTrieNodeFromPeerTask;
+import org.hyperledger.besu.ethereum.eth.manager.task.EthTask;
+import org.hyperledger.besu.ethereum.eth.manager.task.RetryingGetNodeDataFromPeerTask;
+import org.hyperledger.besu.ethereum.eth.sync.DefaultSynchronizer;
+import org.hyperledger.besu.ethereum.trie.CompactEncoding;
+import org.hyperledger.besu.ethereum.worldstate.PeerTrieNodeFinder;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** This class is used to retrieve missing nodes in the trie by querying the peers */
+public class WorldStatePeerTrieNodeFinder implements PeerTrieNodeFinder {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultSynchronizer.class);
+
+  private final Cache<Bytes32, Bytes> foundNodes =
+      CacheBuilder.newBuilder().maximumSize(10_000).expireAfterWrite(5, TimeUnit.MINUTES).build();
+
+  private static final long TIMEOUT_SECONDS = 1;
+
+  final EthContext ethContext;
+  final Blockchain blockchain;
+  final MetricsSystem metricsSystem;
+
+  public WorldStatePeerTrieNodeFinder(
+      final EthContext ethContext, final Blockchain blockchain, final MetricsSystem metricsSystem) {
+    this.ethContext = ethContext;
+    this.blockchain = blockchain;
+    this.metricsSystem = metricsSystem;
+  }
+
+  @Override
+  public Optional<Bytes> getAccountStateTrieNode(final Bytes location, final Bytes32 nodeHash) {
+    Optional<Bytes> ifPresent = Optional.ofNullable(foundNodes.getIfPresent(nodeHash));
```

Comment:
because we can perform block creations as for the rememerBlock without modifying the worldstate. and I don't want to persist the nodes when reading in the trie to avoid conflicts. we can read the trie at any time (during transaction validations, block creations, etc.). If I write at that moment I can take the risk of impacting a block persist that would take place at the same time.

if we do this during a block import. the persist will save the new value and we will no longer need to call this method